### PR TITLE
fix: query timeout kill query directed to incorrect instance

### PR DESCRIPTION
**Pull request by crystall-bitquill**
### Summary

fix: query timeout kill query directed to incorrect instance

### Description

When a query timeout is set and the timeout value has passed, the driver sends a kill query to stop the process through a new connection that was created using the same host as the original connection. When using the reader cluster url, connections are randomly sent to one of the available reader instances. This will occasionally cause errors when the kill query is sent to a different instance where the statement was not executed, resulting in an "Unknown process id" message.

This PR adds a check in the default connection plugin for connections using the reader cluster url. Once the connection is made using the reader cluster url, a new connection is made to a reader instance that is obtained by querying for the current instance name. The original connection using the reader cluster is replaced with the connection to the instance.

Addresses #461

### CODE REVIEW

File: src/test/java/testsuite/integration/container/AuroraMysqlIntegrationTest.java [hunk 1]
```diff
@@ -607,4 +618,16 @@ public void test_PreparedStatementHashCodes() throws SQLException, IOException {
     conn.close();
   }
 
+  @RepeatedTest(100)
```
Comment (sergiyvamz) [inline src/test/java/testsuite/integration/container/AuroraMysqlIntegrationTest.java:621]: Why do we need to run it so many times?

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/DefaultConnectionPlugin.java [hunk 2]
```diff
@@ -100,8 +107,24 @@ public void openInitialConnection(ConnectionUrl connectionUrl) throws SQLExcepti
...
+          rdsUtils.getRdsInstanceHostPattern(mainHostInfo.getHost()).replace("?", connectedHostName);
+      if (!StringUtils.isNullOrEmpty(mainHostInfo.getHostProperties().get(PropertyKey.clusterInstanceHostPattern.getKeyName()))) {
+        instanceEndpoint = mainHostInfo.getHostProperties().get(PropertyKey.clusterInstanceHostPattern.getKeyName())
+                .replace("?", connectedHostName);
+      }
+      final HostInfo instanceHostInfo = ConnectionUtils.createInstanceHostWithProperties(instanceEndpoint, mainHostInfo);
+      final JdbcConnection hostConnection = this.connectionProvider.connect(instanceHostInfo);
+      connection.close();
+      connection = hostConnection;
+    }
```
Comment (sergiyvamz) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/DefaultConnectionPlugin.java:126]: I'd suggest to relocate this logic into a new plugin. In this way, it'd be easier for users to use it when it's needed, or skip it. That gives more flexibility for users who don't experience any issue with query timeout.

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/RdsUtils.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License, version 2.0, for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see
+ * http://www.gnu.org/licenses/gpl-2.0.html.
+ */
+
+package com.mysql.cj.jdbc.ha;
```
Comment (sergiyvamz) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/RdsUtils.java:32]: com.mysql.cj.jdbc.ha.util ?

Comment (sergiyvamz) [reply src/main/user-impl/java/com/mysql/cj/jdbc/ha/RdsUtils.java:32]: And similar suggestion's for the ConnectionUtils class above.

File: src/test/java/testsuite/integration/container/AuroraMysqlIntegrationTest.java (hunk 1 above)
Comment (crystall-bitquill) [reply src/test/java/testsuite/integration/container/AuroraMysqlIntegrationTest.java:621]: When the new connection for the query timeout is made, the reader cluster endpoint doesn't always choose a reader different from the original connection so it's hard to reproduce this issue consistently.

**Review by sergiyvamz — APPROVED** [review]

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/ReaderClusterConnectionPlugin.java [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+ * the GPLv2, the authors of this program hereby grant you an additional
+ * permission to link the program and your derivative works with the
+ * separately licensed software that they have included with the program.
+ *
+ * Without limiting the foregoing grant of rights under the GPLv2 and
+ * additional permission as to separately licensed software, this
+ * program is also subject to the Universal FOSS Exception, version 1.0,
+ * a copy of which can be found along with its FAQ at
+ * http://oss.oracle.com/licenses/universal-foss-exception.
+ *
```
Comment (karenc-bq) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/ReaderClusterConnectionPlugin.java:21]: Part of the license is duplicated.

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/ReaderClusterConnectionPlugin.java [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+         logger,
+         new BasicConnectionProvider());
+  }
+
+  public ReaderClusterConnectionPlugin(
+      ICurrentConnectionProvider currentConnectionProvider,
+      IConnectionPlugin nextPlugin,
+      Log logger,
+      IConnectionProvider connectionProvider) {
+    if (logger == null) {
```
Comment (karenc-bq) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/ReaderClusterConnectionPlugin.java:75]: I don't think logger is used in this file. Do we need to store it?

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/ReaderClusterConnectionPlugin.java [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+        final String connectedHostName = getCurrentlyConnectedInstance(currentConnection);
+        String instanceEndpoint =
+            rdsUtils.getRdsInstanceHostPattern(mainHostInfo.getHost()).replace("?", connectedHostName);
+        if (!StringUtils.isNullOrEmpty(
+              mainHostInfo.getHostProperties().get(PropertyKey.clusterInstanceHostPattern.getKeyName()))) {
+          instanceEndpoint = mainHostInfo.getHostProperties().get(PropertyKey.clusterInstanceHostPattern.getKeyName())
+              .replace("?", connectedHostName);
+        }
+        final HostInfo instanceHostInfo =
+            ConnectionUtils.createInstanceHostWithProperties(instanceEndpoint, mainHostInfo);
```
Comment (karenc-bq) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/ReaderClusterConnectionPlugin.java:120]:
```suggestion
        final String pattern = mainHostInfo.getHostProperties().get(PropertyKey.clusterInstanceHostPattern.getKeyName());
        final String instanceEndpoint = !StringUtils.isNullOrEmpty(pattern) ? pattern : rdsUtils.getRdsInstanceHostPattern(mainHostInfo.getHost());
        final HostInfo instanceHostInfo =
            ConnectionUtils.createInstanceHostWithProperties(
                instanceEndpoint.replace("?", connectedHostName), 
                mainHostInfo);
```

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/ReaderClusterConnectionPlugin.java [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+  @Override
+  public void transactionCompleted() {
+    this.nextPlugin.transactionCompleted();
+  }
+
+  private String getCurrentlyConnectedInstance(JdbcConnection connection) throws SQLException {
+    try (final Statement statement = connection.createStatement()) {
+      final ResultSet rs = statement.executeQuery(GET_INSTANCE_QUERY);
+      rs.next();
+      return rs.getString(1);
```
Comment (karenc-bq) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/ReaderClusterConnectionPlugin.java:146]:
```suggestion
      if (rs.next()) {
        return rs.getString(1);
      }

      return null;
    }
```

**Review by karenc-bq — APPROVED** [review]

**Pull request merged**